blog/2023/06/PSA-dont-use-threadid.md

```diff
@@ -0,0 +1,232 @@
++++
+mintoclevel = 2
+maxtoclevel = 3
+title = "PSA: Thread-local state is no longer recommended"
+authors = "Mason Protter, Valentin Churavy, Ian Butterworth, ..."
+published = "30 June 2023"
+rss_pubdate = Date(2023, 06, 30)
+rss = """PSA: Thread-local state is no longer recommended; Common misconceptions about threadid() and nthreads()"""
++++
+
+
+*Common misconceptions about threadid() and nthreads()*
+
+\toc
+
+
+
+## The buggy pattern
+
+Partially due to the evolving history of our parallel and concurrent interfaces[^historynote], some Julia programmers have been writing *incorrect* parallel code that contains the possibility of race conditions which can give wrong results. This pattern has even been erroneously recommended in [previous official julia blogposts](https://github.com/JuliaLang/www.julialang.org/blob/main/blog/2019/07/multithreading.md#thread-local-state). It is important for the stability and correctness of the ecosystem that these usages are identified and fixed.
+
+The general template that this incorrect code follows is something like the following:
+
+```julia
+using Base.Threads: nthreads, @threads, threadid
+
+states = [some_initial_value for _ in 1:nthreads()]
+@threads for x in some_data
+    tid = threadid()                       
+    old_val = states[tid]
+    new_val = some_operator(old_val, f(x))
+    states[tid] = new_val
+end
+do_something(states)
+```
+
+The above code is **incorrect** because the tasks spawned by `@threads` are  allowed to yield to other tasks during their execution[^yielding]. This means that between reading `old_val` and storing `new_val` in the storage, the task could be paused and a new task running on the same thread with the same `threadid()` could concurrently write to `states[tid]`, causing a race condition and thus work being lost.
+
+This is not actually a problem with multithreading specifically, but really a concurrency problem, and it can be demonstrated even with a single thread. For example:
+
+```
+$ julia --threads=1
+```
+
+```julia
+julia> f(i) = (sleep(0.001); i);
+
+julia> let state = [0], N=100
+           @sync for i ∈ 1:N
+               Threads.@spawn begin
+                   tid = Threads.threadid()  # each task gets `tid = 1`
+                   old_var = state[tid]      # each task reads the current value, which for all is 0 (!) because... 
+                   new_var = old_var + f(i)  # ...the `sleep` in `f` causes all tasks to pause *simultaneously* here (all loop iterations start, but do not yet finish)
+                   state[tid] = new_var      # after being released from the `sleep`, each task sets `state[1]` to `i`
+               end
+           end
+           sum(state), sum(1:N)
+       end
+(100, 5050)
```

Comment:
This wrapping is based on what the preview shows without scrolling
```suggestion
julia> let state = [0], N=100
           @sync for i ∈ 1:N
               Threads.@spawn begin
                   tid = Threads.threadid()  # Each task gets `tid = 1`.
                   old_var = state[tid]      # Each task reads the current value, which for 
                                             # all is 0 (!) because...
                   new_var = old_var + f(i)  # ...the `sleep` in `f` causes all tasks to pause 
                                             # *simultaneously* here (all loop iterations start, 
                                             # but do not yet finish).
                   state[tid] = new_var      # After being released from the `sleep`, each task 
                                             # sets `state[1]` to `i`.
               end
           end
           sum(state), sum(1:N)
       end
(100, 5050)
```

Reply:
Isn't this going  to be device dependant?

Reply:
The suggestion above just fixes it on a regular computer browser view. Beyond that is too difficult to cater for IMO

Reply:
I don't think this should hold up publishing this. Whatever you think best @MasonProtter